### PR TITLE
Use a BasicObject with method_missing for the builder API

### PR DESCRIPTION
**Pull request by timriley**
Using define_singleton_method was bad because interferes with Ruby’s method caching.

I want to call out how I used `respond_to?` here. Because `Builder` is a `BasicObject` subclass now, and because `BasicObject` does not even implement `#respond_to?`, I just implemented that directly and therefore didn't have to bother with `#respond_to_missing?`.

### CODE REVIEW

File: lib/dry/auto_inject/builder.rb [hunk 1]
```diff
@@ -13,13 +13,16 @@ class Builder
       def initialize(container, options = {})
         @container = container
         @strategies = options.fetch(:strategies) { Strategies }
+      end
+
+      def method_missing(name, *args, &block)
+        return super unless strategies.key?(name)
+
+        Injector.new(container, strategies[name])
+      end
```
Comment (solnic) [inline lib/dry/auto_inject/builder.rb]: I know this is minor stuff but a simple if/else is always easier to parse with eyes than `unless` with return so:

``` ruby
if strategies.key?(name)
  Injector.new(container, strategies[name])
else
  super
end
```

Comment (solnic) [inline lib/dry/auto_inject/builder.rb]: oh and `method_missing` should be private

**Comment by solnic** [comment]
It would be nice to actually have specs for method_missing behavior, it's always *so easy* to mess it up by an accident and libs with messed up method missing behavior may cause weird issues :)

File: lib/dry/auto_inject/builder.rb (hunk 1 above)
Comment (timriley) [inline lib/dry/auto_inject/builder.rb]: Yeah, you're right @solnic, that if/else is much clearer for a method with only 2 single-line pathways :)

**Pull request merged**